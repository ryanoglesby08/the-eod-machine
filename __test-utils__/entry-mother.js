const pick = require('./pick')

const defaultEntry = {
  author: 'Some author',
  category: 'Test category',
  content: 'some content',
}

const someEntryInput = (overrides = {}) => {
  return Object.assign(
    {},
    pick(defaultEntry, ['category', 'content']),
    pick(overrides, ['category', 'content'])
  )
}

const anAuthoredEntry = (overrides = {}) => {
  return Object.assign({}, defaultEntry, overrides)
}

const someEntryInputAndAuthoredEntry = (overrides = {}) => {
  return {
    entryInput: someEntryInput(overrides),
    authoredEntry: anAuthoredEntry(overrides),
  }
}

module.exports = {
  someEntryInput,
  anAuthoredEntry,
  someEntryInputAndAuthoredEntry,
}
