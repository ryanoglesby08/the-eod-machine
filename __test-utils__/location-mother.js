const pick = require('./pick')

const baseLocation = {
  name: 'The location',
  timeZone: 'The time zone',
  eodTime: 'The EOD time',
}

const createLocationMother = relevantFields => (overrides = {}) => {
  return Object.assign(
    {},
    pick(baseLocation, relevantFields),
    pick(overrides, relevantFields)
  )
}

const aLocation = (overrides = {}) => {
  return Object.assign({}, baseLocation, overrides)
}

module.exports = { baseLocation, aLocation, createLocationMother }
