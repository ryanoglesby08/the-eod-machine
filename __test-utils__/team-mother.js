const pick = require('./pick')
const { anAuthoredEntry } = require('./entry-mother')

const defaultTeam = {
  _id: 'team-0',
  name: 'The team',
  mailingList: ['team@example.com'],
  currentEod: [anAuthoredEntry()],
}

const aTeamWithItsEod = (overrides = {}) => {
  return Object.assign({}, defaultTeam, overrides)
}

const someTeamInput = (overrides = {}) => {
  return Object.assign(
    {},
    pick(defaultTeam, ['name', 'mailingList']),
    overrides
  )
}

module.exports = { aTeamWithItsEod, someTeamInput }
