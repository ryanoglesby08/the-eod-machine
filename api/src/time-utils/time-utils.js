import { zonedTimeToUtc } from 'date-fns-tz'

export const to24HourTime = ampmTime => {
  const [time, ampm] = ampmTime.split(' ')
  let [hours, minutes] = time.split(':')

  hours = parseInt(hours)
  minutes = parseInt(minutes)

  if (hours === 12 && ampm === 'AM') {
    hours = 0
  }

  if (hours !== 12 && ampm === 'PM') {
    hours += 12
  }

  return { hours, minutes }
}

export const convertLocalTimeToUtcTime = (localTime, localTimeZone) => {
  const { hours, minutes } = to24HourTime(localTime)

  const localDate = new Date()
  localDate.setHours(hours, minutes, 0, 0)

  const utcDate = zonedTimeToUtc(localDate, localTimeZone)
  const utcTime = utcDate.toLocaleTimeString('en-US', {
    timeZone: 'UTC',
    hour: 'numeric',
    minute: '2-digit',
    hour12: false,
  })

  return utcTime
}
