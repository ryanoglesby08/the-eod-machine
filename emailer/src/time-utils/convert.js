import { zonedTimeToUtc } from 'date-fns-tz'

export const to24HourTime = ampmTime => {
  const [time, ampm] = ampmTime.split(' ')
  let [hours, minutes] = time.split(':').map(t => parseInt(t))

  if (hours === 12 && ampm === 'AM') {
    hours = 0
  }

  if (hours !== 12 && ampm === 'PM') {
    hours += 12
  }

  return { hours, minutes }
}

export const convertLocalTimeToUtcTime = (localTime, localTimeZone) => {
  const { hours, minutes } = to24HourTime(localTime)

  const localDate = new Date()
  localDate.setHours(hours, minutes, 0, 0)

  const utcDate = zonedTimeToUtc(localDate, localTimeZone)
  const utcTime = utcDate.toLocaleTimeString('en-US', {
    timeZone: 'UTC',
    hour: 'numeric',
    minute: '2-digit',
    hour12: false,
  })

  return utcTime
}

export const roundToNearestHalfHour = utcTimeString => {
  const [hours, minutes] = utcTimeString.split(':').map(t => parseInt(t))

  const roundedMinutes = Math.round(minutes / 30) * 30

  const time = new Date()
  time.setHours(hours, roundedMinutes, 0, 0)

  return time.toLocaleTimeString('en-US', {
    hour: 'numeric',
    minute: '2-digit',
    hour12: false,
  })
}
