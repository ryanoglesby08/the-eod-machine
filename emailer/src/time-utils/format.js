export const toShortDate = date =>
  date.toLocaleString('en-US', {
    month: 'short',
    day: '2-digit',
  })

export const toLongDate = date =>
  date.toLocaleString('en-US', {
    month: 'short',
    day: '2-digit',
    year: 'numeric',
  })
