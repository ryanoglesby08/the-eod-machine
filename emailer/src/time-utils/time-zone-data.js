import timeZonesData from 'timezones.json'

const timeZonesDataByValue = timeZonesData.reduce(
  (accumulator, currentTimeZoneObject) => {
    return {
      ...accumulator,
      [currentTimeZoneObject.value]: currentTimeZoneObject,
    }
  },
  {}
)

export const getUtcTimezone = timeZone => timeZonesDataByValue[timeZone].utc[0]

export default timeZonesDataByValue
