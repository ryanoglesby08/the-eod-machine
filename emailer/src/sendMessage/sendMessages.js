import gql from 'graphql-tag'
import { utcToZonedTime } from 'date-fns-tz'

import apiClient from '../apiClient'

import { getUtcTimezone } from '../time-utils/time-zone-data'
import { convertLocalTimeToUtcTime } from '../time-utils/convert'
import { toShortDate } from '../time-utils/format'

import { renderHtmlMessage } from '../multipartMessage/HtmlMessage/HtmlMessage'
import textMessage from '../multipartMessage/textMessage/textMessage'

export const GET_TEAMS_READY_FOR_EOD_DELIVERY = gql`
  query GetTeamsReadyForEodDelivery($currentTimeUtc: String!) {
    teamsReadyForAnEodDelivery(currentTimeUtc: $currentTimeUtc) {
      name
      mailingList
      currentEod {
        author
        category
        content
      }
      locations {
        name
        timeZone
        eodTime
      }
    }
  }
`

const getLocationAtEod = (locations, currentTimeUtc) => {
  return locations.find(({ eodTime, timeZone }) => {
    const eodTimeUtc = convertLocalTimeToUtcTime(
      eodTime,
      getUtcTimezone(timeZone)
    )

    return eodTimeUtc === currentTimeUtc
  })
}

const subject = (team, eodLocation, eodLocationDate) =>
  [`[EOD] ${team.name}`, eodLocation.name, toShortDate(eodLocationDate)].join(
    ' | '
  )

const sendMessages = async (transporter, currentDate, currentTimeUtc) => {
  const {
    data: { teams },
  } = await apiClient.query({ query: GET_TEAMS_READY_FOR_EOD_DELIVERY })

  return await Promise.all(
    teams.map(async team => {
      const eodLocation = getLocationAtEod(team.locations, currentTimeUtc)
      const eodLocationDate = utcToZonedTime(
        currentDate,
        getUtcTimezone(eodLocation.timeZone)
      )

      return await transporter.sendMail({
        from: {
          address: 'eod-machine@theeodmachine.local',
          name: 'EOD Machine',
        },
        to: team.mailingList,
        subject: subject(team, eodLocation, eodLocationDate),
        text: textMessage(team.currentEod, eodLocation, eodLocationDate),
        html: renderHtmlMessage(team.currentEod, eodLocation, eodLocationDate),
      })
    })
  )
}

export default sendMessages
