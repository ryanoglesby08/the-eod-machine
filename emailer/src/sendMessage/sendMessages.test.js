import nodemailer from 'nodemailer'

import createMother from '../../../__test-utils__/graphql-query-mother'
import { baseTeam } from '../../../__test-utils__/team-mother'
import { createLocationMother } from '../../../__test-utils__/location-mother'

import stubApiClient from '../apiClient'

import { convertLocalTimeToUtcTime } from '../time-utils/convert'
import sendMessages, { GET_TEAMS_READY_FOR_EOD_DELIVERY } from './sendMessages'

jest.mock('../apiClient')

const aTeamWithItsEod = createMother(GET_TEAMS_READY_FOR_EOD_DELIVERY, baseTeam)
const aLocation = createLocationMother(['name', 'timeZone', 'eodTime'])

let transporter

beforeEach(() => {
  transporter = nodemailer.createTransport({
    jsonTransport: true,
  })
})

it('sends a message to teams that are ready for an EOD', async () => {
  const team = aTeamWithItsEod({
    name: 'The Team',
    mailingList: ['team1@example.com', 'team2@example.com'],
    locations: [
      aLocation({
        name: 'Denver',
        timeZone: 'Mountain Standard Time',
        eodTime: '5:00 PM',
      }),
      aLocation({
        name: 'New York',
        timeZone: 'Eastern Standard Time',
        eodTime: '8:00 PM',
      }),
    ],
  })

  stubApiClient.__stubQuery(GET_TEAMS_READY_FOR_EOD_DELIVERY, {
    teams: [team],
  })

  const currentDate = new Date(2019, 1, 3, 12) // Feb 3, 2019 @ 12pm UTC
  const currentTimeUtc = convertLocalTimeToUtcTime('5:00 PM', 'America/Denver')

  const messages = await sendMessages(
    // TODO: don't pass transporter as a param, just use module mocking
    transporter,
    currentDate,
    currentTimeUtc
  )

  expect(messages).toHaveLength(1)

  const eodMessage = JSON.parse(messages[0].message)

  expect(eodMessage).toHaveProperty('from', {
    address: 'eod-machine@theeodmachine.local',
    name: 'EOD Machine',
  })
  expect(eodMessage).toHaveProperty('to', [
    { address: 'team1@example.com', name: '' },
    { address: 'team2@example.com', name: '' },
  ])
  // TODO: Test for reply to field
  expect(eodMessage).toHaveProperty(
    'subject',
    '[EOD] The Team | Denver | Feb 03'
  )

  expect(eodMessage.text).toMatchSnapshot()
  expect(eodMessage.html).toMatchSnapshot()
})

// TODO: Next test
//it('rounds to the nearest half hour to account for small time differences when the emailer runs', () => {})
// round to nearest half hour => (Math.round(<current minutes> / 30) * 30) % 60
