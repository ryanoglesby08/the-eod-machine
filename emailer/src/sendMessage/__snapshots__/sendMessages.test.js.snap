// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sends a message to teams that are ready for an EOD 1`] = `
"
EOD updates from Denver for Feb 03, 2019

---------------------------------------------------------------------
Test category

* some content (Some author)


- Delivered by The EOD Machine
"
`;

exports[`sends a message to teams that are ready for an EOD 2`] = `"<h1>EOD updates from Denver for Feb 03, 2019</h1><div><h2>Test category</h2><ul><li>some content (Some author)</li></ul></div><div>Delivered by The EOD Machine</div>"`;
